internal/command/views/json/diagnostic.go

```diff
@@ -416,6 +430,204 @@ func NewDiagnostic(diag tfdiags.Diagnostic, sources map[string][]byte) *Diagnost
 	return diagnostic
 }
 
+// Each line in the out is prepended with this prefix.
+var ttyOutPrefix = "    | "
+
+// marshalValue marshals a cty.Value to a JSON string, or returns a fallback
+// string if any failure is encountered. The marshalled JSON string is pretty
+// printed with a prefix for each line.
+func marshalValue(val cty.Value, fallback string) string {
+	switch {
+	case val.Type().IsPrimitiveType():
+		return fallback
+	default:
+		jsonVal, err := ctyjson.Marshal(val, val.Type())
+		if err != nil {
+			return ""
+		}
+		jsonVal = append(jsonVal, '\n')
+		var str bytes.Buffer
+		if err := json.Indent(&str, jsonVal, ttyOutPrefix, "  "); err != nil {
+			return fallback
+		}
+		return str.String()
+	}
+}
+
+// diffBinaryFailedRunDiagnostic is used to diff the two values of a binary expression
+// in a failed run diagnostic. More often than not, run assert expressions are binary
+// expressions, i.e., `a == b`. If the values are of different types, we don't diff them.
+func diffBinaryFailedRunDiagnostic(ctx *hcl.EvalContext, expr *hclsyntax.BinaryOpExpr) DiagnosticExpressionValue {
+	// The expression has already been evaluated and failed, so we can ignore the diags here.
+	lhs, _ := expr.LHS.Value(ctx)
+	rhs, _ := expr.RHS.Value(ctx)
+
+	// The types do not match. We don't need to diff them?
```

Comment:
I don't think we need to display a diff when the types do not match. The diff may end up being unreadable, and there is little value in it, especially because we are already showing the full objects involved in the operation

Reply:
contrived example of a diff of distinct types

```
│ var.tr1 is [
            |   "foo",
            |   "bar"
            | ]
        
            │ var.tr2 is "foo"
            │ ~diff: 
            | - [
            | + "foo"
            | - "foo",
            | - "bar"
            | - ]
            | - 
```

Reply:
While we allow some comparison to pass in simple primitives (true == "true"), simply ignoring diff when types do not match seem sufficient

Reply:
another edge case is when we have objects of different types, but whose string representation is the same, e.g
```
{
  "iops" = tonumber(null)
  "size" = 60
}

// AND

{
  iops = null
  size = 60
}
```

The diff for this would indicate no visible difference. In that case, perhaps we should just tell that the values do not have the same type (while showing both objects with no visible difference.

Reply:
626a9f4929